src/contract/superblock.cpp

```diff
@@ -0,0 +1,91 @@
+#include "superblock.h"
+#include "uint256.h"
+#include "util.h"
+#include "main.h"
+#include "compat/endian.h"
+
+std::string ExtractValue(std::string data, std::string delimiter, int pos);
+
+namespace
+{
+    struct BinaryResearcher
+    {
+        std::array<unsigned char, 16> cpid;
+        int16_t magnitude;
+    };
+}
+
+std::string UnpackBinarySuperblock(std::string sBlock)
+{
+    // 12-21-2015: R HALFORD: If the block is not binary, return the legacy format for backward compatibility
+    std::string sBinary = ExtractXML(sBlock,"<BINARY>","</BINARY>");
+    if (sBinary.empty()) return sBlock;
+
+    std::ostringstream stream;
+    stream << "<AVERAGES>" << ExtractXML(sBlock,"<AVERAGES>","</AVERAGES>") << "</AVERAGES>"
+           << "<QUOTES>" << ExtractXML(sBlock,"<QUOTES>","</QUOTES>") << "</QUOTES>"
+           << "<MAGNITUDES>";
+
+    // Binary data support structure:
+    // Each CPID consumes 16 bytes and 2 bytes for magnitude: (Except CPIDs with zero magnitude - the count of those is stored in XML node <ZERO> to save space)
+    // 1234567890123456MM
+    // MM = Magnitude stored as 2 bytes
+    // No delimiter between CPIDs, Step Rate = 18.
+    // CPID and magnitude are stored in big endian.
+    for (unsigned int x = 0; x < sBinary.length(); x += 18)
+    {
+        if(sBinary.length() - x < 18)
+            break;
+
+        const BinaryResearcher* researcher = reinterpret_cast<const BinaryResearcher*>(sBinary.data() + x);
```

Comment:
This is endian unsafe and alignment unsafe.

Reply:
It is endian safe (see calls to be16toh and htobe16) but not alignment safe. I'll see if we can add static asserts for it so we can deal with it as new platforms come along. Good catch.